### DISCUSSION TITLE
Function overloading in macros

### DISCUSSION THREAD

**Discussion by ristomatti** [houmain/keymapper#176] [Ideas]
@houmain Would it be a lot of work to support the same macro with varying amount of parameters? Example:

```ini
Hold = $0{150ms}
Hold = $0{mul[150, $1]ms}
```

The idea came from the fact I've found my macros `Tap` and `Hold` great ways to make the config more readable and to allow fine tuning the value for all the mappings in one place. But in a few situations I need the value to be longer, but I couldn't come up with a good name. It then occurred to me that using a multiplier would be a even better to have the value relative to the most often needed one.

Currently I've just used `HoldX`, but I remember having a similar need on some macros that execute commands earlier on. For instance it would be nice to have this

```ini
log = $(echo "$1" >> $LOG_FILE)
log = $(pastel --force-color paint "$1" "$0" >> $LOG_FILE)
```

...rather than this:

```ini
log = $(echo "$1" >> $LOG_FILE)
logColor = $(pastel --force-color paint "$1" "$0" >> $LOG_FILE)
```

What do you think?

**Comment by houmain** [comment]
> Would it be a lot of work to support the same macro with varying amount of parameters?

Not only that, it would be against the design. It is currently allowed to redefine macros, like:

```
macro = X
A >> macro
macro = Y
B >> macro
```

So, it would be hard to distinguish between a redefinition and a definition of an overload.

I am thinking about adding a builtin macro `default`, which would return the second parameter when the first is empty. So this would work:

```
Hold = $0{mul[150, default[$1, 1]]ms}
```

I am still thinking about a nice solution for your second sample.

**Reply by ristomatti** [reply]
I think passing in a default value makes the most sense. Just go with it and forget the second example. 😄

**Reply by houmain** [reply]
I added the builtin macro `default`.

And also a `$#`, which is replaced with the argument count.
So now it is at least possible to overload:
```
log1 = $(echo "$0" >> $LOG_FILE)
log2 = $(pastel --force-color paint "$1" "$0" >> $LOG_FILE)
logN = log$0
log = logN[$#][$0, $1]
```
logN returns log1 or log2, which is then called with the second argument list.
Beauty is in the eye of the beholder.

**Reply by ristomatti** [reply]
I like the simplicity of it. It makes me worry less about regression. Besides, it's all going to be swept under the carpet behind `@include`. It's the beauty of how it looks within mappings that matters (at least for me that is). :grin: 

I noticed some commits pushed, do you want me  to give it a try it before a release?

**Reply by ristomatti** [reply]
Aside: I haven't pushed any commits to my config repo in a while. It's because I've managed to litter my config with too many changes, and I hate vague commits consisting of several unrelated changes.

One of the changes was removing a QMK based tap/hold mapping on my most used mod key between split base bars with the goal of reducing the cognitive overhead and simplifying Keymapper config. I'm not 100% happy with the result though. I mapped the thumb key to `Exec` thinking it would make it easier to customize the behavior within the initial stage. It turned out to introduce more complexity within my Keymapper config I anticipated... Since I primary use it as `Alt`/`Escape`, I probably should have made either of these two. :thinking:

**Reply by ristomatti** [reply]
I have come up with some other perhaps more interesting changes, mostly tied to i3wm though. Perhaps I'd better try committing out the pending changes before it's too late to avoid the "Update config" commit. :roll_eyes:

**Reply by ristomatti** [reply]
> And also a $#, which is replaced with the argument count.

Hmm. Could you reconsider using some other character than`#`? It makes sense given the meaning, but it will mess up the syntax highlighting with the only languages I've got decent results. This is with Starlark, the most `keymapper.conf` "compatible" one I've come across:

![image](https://github.com/user-attachments/assets/0cec2e58-8c94-49d8-999d-856a7a7e509f)

**Reply by ristomatti** [reply]
Before I discovered Starlark worked so well for this purpose, I went through using Bash, INI and Perl but all of them started falling apart with some of Keymapper's config syntax. 

I have even contemplated on trying to write Tree-sitter highlighting queries for Keymapper but the task has so far felt too daunting. I don't even know if it'd be doable without writing a parser, but the highlighting query syntax is already difficult for me to wrap my head around. I might be able to write a TextMate higlighter but unfortunately Helix doesn't support it...

It doesn't help at all Tree-sitter grammars are [written in Scheme](https://github.com/tree-sitter-grammars/tree-sitter-starlark/blob/master/queries/) and I've only worked with languages in the [C-family](https://en.wikipedia.org/wiki/List_of_C-family_programming_languages).

**Reply by ristomatti** [reply]
Perhaps `%`?

![image](https://github.com/user-attachments/assets/8f476096-edb1-4136-85ba-6641d889957e)

**Reply by ristomatti** [reply]
Related: https://www.quora.com/Is-the-number-sign-internationally-recognized-to-mean-number

**Reply by houmain** [reply]
> I noticed some commits pushed, do you want me to give it a try it before a release?

Early feedback is always very welcome!

> Could you reconsider using some other character than `#`?

I did not think about that. I just copied the syntax from shell scripts. But since `#` has indeed been problematic I changed it to `$$` now.

**Reply by ristomatti** [reply]
Sorry to respond late rather than early, just haven't found the time earlier on. Both `default` and `$$` work as expected. I also noticed you've added `apply` which seems brilliant. Great work once again!

I might have both time and inspiration update my config repo this evening. If I do, I will definitely include usages of both `default` and `$$`. I think there  was also some mappings that  could be simplified with `apply`. :thinking: 

> Not only that, it would be against the design. It is currently allowed to redefine macros, like:
> 
> ```
> macro = X
> A >> macro
> macro = Y
> B >> macro
> ```

I had missed the gist of this comment. Interesting! It hasn't occurred for me to test this earlier. I was under the impression the last definition would apply to the full config, but it's just the order in which they appear in the config that matters?

**Reply by houmain** [reply]
> I might have both time and inspiration update my config repo this evening.

I am looking forward to see how you put them to good use!

> but it's just the order in which they appear in the config that matters?

Yes, macros are immediately evaluated when a configuration line is parsed.

**Comment by ristomatti** [comment]
Both use cases can now be implemented thanks to the new `default` macro and the new special variable `$$` introduced in [version 4.9.0](https://github.com/houmain/keymapper/releases/tag/4.9.0).

```starlark
# "Hold" use case
Hold = $0{mul[130, default[$1, 1]]ms}

# "log" use case
switch = $0$1

_log1  = $(echo "$0" >> $LOG_FILE)
_log2  = $(pastel --force-color paint "$1" "$0" >> $LOG_FILE)

log    = switch["_log", $$]["$0", "$1"]
```
